### PR TITLE
[lldb][test] Enable static linking with libcxx for import-std-module tests

### PR DESCRIPTION
**Pull request by dzhidzhoev**
--whole-archive and --allow-multiple-definition options has been added to linker flags of these import-std-module tests in order to make them pass with libcxx static linking enabled.

Darwin has been excluded since it doesn't seem to be built in this configuration.

Also, some tests were switched from system stdlib to libcxx since the problem described in https://reviews.llvm.org/D139361 seems to be fixed.

### CODE REVIEW

File: lldb/test/API/commands/expression/import-std-module/deque-dbg-info-content/Makefile [hunk 1]
```diff
@@ -1,5 +1,8 @@
-# FIXME: once the expression evaluator can handle std libraries with debug
-# info, change this to USE_LIBCPP=1
-USE_SYSTEM_STDLIB := 1
```
Comment (Michael137) [inline lldb/test/API/commands/expression/import-std-module/deque-dbg-info-content/Makefile:3]: Is this related to the linker flag changes in any way? We rely on this on the macOS bots, so it doesn't sound like the underlying issue the FIXME referred to was resolved. If it's an unrelated change, I'd prefer opening a separate PR for these

File: lldb/test/API/commands/expression/import-std-module/array/Makefile [hunk 2]
```diff
@@ -1,3 +1,8 @@
 USE_LIBCPP := 1
 CXX_SOURCES := main.cpp
+
+ifneq ($(OS),Darwin)
+	LD_EXTRAS := -Xlinker --whole-archive -Xlinker --allow-multiple-definition
```
Comment (Michael137) [inline lldb/test/API/commands/expression/import-std-module/array/Makefile:5]: Can you elaborate on why this is necessary? Example test failures would be great

**Comment by slydiman** [comment]
It seems #106885 fixed the issue for 
`deque-dbg-info-content/TestDbgInfoContentDequeFromStdModule.py` 
and 
`list-dbg-info-content/TestDbgInfoContentListFromStdModule.py`, 
but not for `vector-dbg-info-content/TestDbgInfoContentVectorFromStdModule.py`.

The following 3 tests don't let the [buildbot](https://lab.llvm.org/staging/#/builders/195/builds/4450) turn green
array/TestArrayFromStdModule.py
vector-of-vectors/TestVectorOfVectorsFromStdModule.py
vector-dbg-info-content/TestDbgInfoContentVectorFromStdModule.py

**Comment by Michael137** [comment]
Can you add in the PR description why the `--whole-archive and --allow-multiple-definition` fixes the issue? And paste the log of the CI test that's failing?

**Comment by slydiman** [comment]
> Or is @slydiman implying this PR isn't needed anymore?

This patch should fix the following tests:
array/TestArrayFromStdModule.py
vector-of-vectors/TestVectorOfVectorsFromStdModule.py
vector-dbg-info-content/TestDbgInfoContentVectorFromStdModule.py

See the buildbot https://lab.llvm.org/staging/#/builders/195/builds/4464
Note it is the first buildbot for cross build and it may turn green for the first time after fixing these tests.

But probably it is possible to update this patch to keep `deque-dbg-info-content/Makefile` and `list-dbg-info-content/Makefile` unchanged.

**Comment by labath** [comment]
I doubt those linker flags are going to work on windows. It might be more "portable" to ensure the symbols are present by using them from inside the program.

**Comment by slydiman** [comment]
Those linker flags work on Windows. 
Note I tried to add the usage of a.at(0) to main.cpp and it did not help.
Please see the alternative patch #112485.

**Comment by labath** [comment]
I guess it will with clang+lld (which is the only way we can run the test suite now), though I still find this suboptimal -- in particular, I think the multiple definition thingy plants us firmly into undefined behavior territory.

Did you check that `at(0)` actually causes the necessary symbols to be pulled into the binary? If yes, why can't we find them? If not, why not?

**Comment by Michael137** [comment]
> I guess it will with clang+lld (which is the only way we can run the test suite now), though I still find this suboptimal -- in particular, I think the multiple definition thingy plants us firmly into undefined behavior territory.
> 
> Did you check that `at(0)` actually causes the necessary symbols to be pulled into the binary? If yes, why can't we find them? If not, why not?

Agree with Pavel here, it would be good to figure out what exactly is going wrong in this test.

FWIW, we don't actually want to emit `at` into the debug-info. This test specifically tests the case where we don't have the debug-info for a function, but are able to find and call it using the `std` Clang module. IIRC from the last time we saw this, the problem only occurs when debug symbols for libc++ are available (https://reviews.llvm.org/D139361).

Since this is quite tricky to debug, and has been a known issue for a while and (i think) only happens when we built libc++ with debug symbols, I'd be fine with skipping the test on Linux until we figure out what the root cause is.

**Comment by slydiman** [comment]
> Since this is quite tricky to debug, and has been a known issue for a while and (i think) only happens when we built libc++ with debug symbols, I'd be fine with skipping the test on Linux until we figure out what the root cause is.

Agree. Please see #112530.

**Comment by dzhidzhoev** [comment]
Closed in favor of https://github.com/llvm/llvm-project/pull/118986
